Implemented fractional part function

leosartaj opened the pull request:
```
>>> from sympy import frac, Rational, floor
>>> from sympy.abc import x, n
>>> frac(x)
frac(x)
```

returns zero for integer arguments

```
>>> frac(n)
0
>>> frac(Rational(8, 5))
Rational(3, 5)
```

rewrite as floor

```
>>> frac(x).rewrite(floor)
x - floor(x)
```

closes #8674 
@jksuom @debugger22

leosartaj commented:
Currently it uses the definition ``frac(x) = x - floor(x)``. For extending it to negative integers I found different definitions [here](http://en.wikipedia.org/wiki/Fractional_part).
 
Taking wolfram as base, we can define it as an odd function. Moreover, fractional part of complex numbers can be defined as, ``frac(x + I*y) = frac(x) + I*frac(y)``.

Also it would be nice to have 

```
>>> frac(x) < 1
True
>>> frac(x) > -1
True
```

Though, am not sure how to do it in sympy.

toolforger commented:
There is a known_facts array that holds this kind of assertions.
I don't know how it is used though.

leosartaj commented:
@jksuom if you could have a look.

jksuom commented:
There does not seem to exist any standard definition of a fractional part function. Even for positive real numbers it could be ``x - floor(x)`` or ``x % 1``. I am not able to decide which one to choose as ``frac(x)``. Such expressions are so simple that it would be possible to leave the choice to the user for now.

toolforger commented:
I wouldn't want to bother users with simple decisions.

leosartaj commented:
I think, we should decide on a single definition. This could be clearly stated in the documentation. I prefer wolfram's [definition](http://en.wikipedia.org/wiki/Fractional_part). Only reason being, that wolfram uses it, hence, it's tried and tested, therefore works well. 

Other issue is, whether it should be extended for complex arguments or not. This really confuses me, wolfram does it. But mathematically I can't get my head around the fractional part of something like ``a + i*b``.

leosartaj commented:
What do you think? @jksuom

jksuom commented:
Perhaps my response above was too concise. I shall try to clarify.

I think there are two aspects in the definition of a function.

1. Value
2. Implementation

All definitions agree on the value of ``frac(x)`` for positive real ``x``. That should be no problem here.

For the implementation there are (at least) two possibilities, ``x - floor(x)`` and ``x % 1`` (which seem to give the same value even for negative real ``x``).

I believe the implementation should be chosen to be as efficient as possible. That should be investigated. Efficiency may even depend on the context, in which case the end user might want to decide (as e.g. [here](https://github.com/cbm755/sympy/blob/master/sympy/simplify/hyperexpand.py#L92)).

toolforger commented:
I wouldn't spend too much time worrying about efficiency. Quote Edsger W. Dijkstra: "Premature optimization is the root of all evil."
Usually you'll find that differences of that kind aren't even measurable. Optimizations like the link @jksuom provided are there to prevent heavy SymPy machinery from being uselessly applied; choosing between `x - floor(x)` and `x % 1` does not change one bit about that.

One rule of thumb to go by is "minimize the number of function calls". Because more complicated expressions make it harder to reason about an expression (and this can save big on CPU cycles), so this speaks in favor of `x % 1`.
On the other hand, "minimize the number of functions being called at the bottom", because that makes it easier to normalize expressions and see whether things cancel each other out; `%` is a pretty rare operator to use, so it's unlikely that SymPy code knows how to transform that well, but it does know enough about algebraic laws to cancel out `floor(x)` and `- floor(x)` if they happen to appear in the same sum term.
In other words, there's no clear choice - but if you make it a user choice, then users will have to know which of the two arguments carries more weight. Most likely it will just be useless though, and a distraction since it's yet another knob to think about and choose whether to try it or not (this becomes really nasty if knobs become relevant only if other knobs are turned as well). Essentially, SymPy itself is usuall in a better position to decide what's best.

So: Wait until an actual use case arises. Then decide what do to. If anything should be done at all.
If you feel unsure, throw a die.

jksuom commented:
> So: Wait until an actual use case arises. Then decide what do to. If anything should be done at all.

I agree.

leosartaj commented:
I have extend ``frac`` for negative arguments using the [definition](http://mathworld.wolfram.com/FractionalPart.html) wolfram(and even maple) uses. I preferred this due to two reasons:
* I like the fact, that int(x) + frac(x) = x.
* As discussed above, it all boils down to use cases. Since, it's wolfram that uses this definition. I guess it works well.

@jksuom @toolforger

toolforger commented:
Am 11.05.2015 um 23:57 schrieb Sartaj Singh:
> I have extend ``frac`` for negative arguments using the [definition](http://mathworld.wolfram.com/FractionalPart.html) wolfram(and even maple) uses.
 >
 > I preferred this due to two reasons:
 >
> * I like the fact, that int(x) + frac(x) = x.

The other definition would be `floor(x) + frac(x) = x`.
I.e. having a nice formula is possible whatever you do.

> * As discussed above, it all boils down to use cases. Since, it's wolfram that uses this definition. I guess it works well.

Wolfram offers both functions: `FractionalPart` = `x - int(x)`, and 
`SawtoothWave` = `x - floor(x)`.
Maybe that's the way to go, though I wouldn't want to suddenly use 
geometric terminology like `SawtoothWave`.

Another heads-up: our `floor` differs from Wolfram's `floor` for complex 
numbers. I.e. if we really want to be roughly analogous to Wolfram, we 
need to change the meaning of `floor` (and `ceiling` I suppose).
Such a change would require a discussion in the mailing list and a 
separate PR though.

leosartaj commented:
Here by int(x) I meant integer part of x. I have used the odd function definition for ``frac``.

Further, I have extended functionality to complex arguments.
```
>>> r = Symbol('r', real=True)
>>> frac(I*r)
I*frac(r)
```

leosartaj commented:
@jksuom @toolforger

toolforger commented:
`int` extends to the complex plane, but it does it differently than Wolfram.
So you can have either a `frac` that's compatible with Wolfram, or have `z = int(z) + frac(z)`.
So I think complex plane extension for `int`, `frac`, `ceil` and `floor` need to be discussed on the mailing list first.
BTW no need to ping unless they haven't responded in a while.

leosartaj commented:
floor, ceiling already supports this kind of functionality

```
>>> r = Symbol('r', real=True)
>>> floor(I*r)
I*floor(r)
>>> ceiling(I*r)
I*ceiling(r)
```

toolforger commented:
Then I guess it was `int`.

leosartaj commented:
If you could please have a look @jksuom

jksuom commented:
I have been trying to recall cases where the fractional part has been actually used. I think that they all have been of the type ``x - floor(x)`` even for negative ``x``. So that is what I would prefer, either in that form or as ``x % 1`` (or perhaps rather ``Mod(x, 1)``).

leosartaj commented:
In school, I have also been used to this definition. Very well then, I will update.

jksuom commented:
Perhaps you could test with some data if there is any significant difference in performance between the two implementations.

leosartaj commented:
Some results

```
>>> from random import uniform
>>> %timeit (frac(uniform(x, x+1)) for x in range(10000, 100000)) # x - floor(x)
1000 loops, best of 3: 696 µs per loop
>>> %timeit (frac(uniform(x, x+1)) for x in range(10000, 100000)) # Mod(x, 1)
1000 loops, best of 3: 695 µs per loop
```

So, the results are similar. I would prefer to go with ``x - floor(x)`` just beacuse it's more familiar and standard.

leosartaj commented:
I have update ``frac`` to use the definition ``x - floor(x)`` for negative/positive numbers. @jksuom

jksuom commented on sympy/functions/elementary/integers.py:
```diff
@@ -220,3 +220,72 @@ def __ge__(self, other):
+        from sympy import im
+        if arg.is_integer:
+            return S.Zero
+        if arg.is_imaginary or (S.ImaginaryUnit*arg).is_real:
```
Is it necessary to have two conditions? In which case is the latter one needed?

jksuom replied:
I also wonder why it should be necessary to deal with a pure imaginary argument in a separate section. I have understood that floor could be used for all complex arguments.

leosartaj replied:
Examples

```
>>> a = sympify(2j)
# frac(a)
>>> a.is_imaginary # works good
True
>>> r = Symbol('r', real=True)
# frac(I*r) -> I*frac(r)
>>> (I*r).is_imaginary
None
>>> (I*(I*r)).is_real
True
```

Also, this line is actually L25, and works well in the case of ``floor`` and ``ceiling``. So, I picked it up without much thought.

leosartaj replied:
I kept ``floor`` related expression inside ``is_number`` block only, because

```
>>> Symbol('x', real=True)
>>> frac(x)
x - floor(x) # not so great, frac(x) is better
```

while, complex simplification should be done in all cases.

jksuom replied:
The failure of ``(I*r).is_imaginary`` seems to be caused by the possibility of ``r == 0``. So perhaps the other test is needed as long as #7649 remains open.

leosartaj replied:
I will update the code with a comment.

jksuom replied:
There might be other ways of avoiding the unevaluated ``x - floor(x)``. Have you tested code of this type:
```
n = floor(x)
if not n.has(floor):
    return x - n
```
If that works, then perhaps the is_imaginary block could be omitted.

leosartaj replied:
Well, I guess this will cause some issues.

Currently
```
>>> r = Symbol('r', real=True)
>>> frac(I*r)
I*frac(r) # expected
```
Now,
```
>>> floor(I*r)
I*floor(r)
>>> frac(I*r)
frac(I*r) # if block did not execute
```

This does avoid unevaluated ``x - floor(x)`` but leaves ``frac`` unevaluated as well.

jksuom replied:
I still find it disturbing that only pure imaginary arguments are treated differently. Otherwise the result seems to remain unevaluated with arguments containing symbols.
```
>>> r = Symbol('r', real=True)
>>> frac(r*I)
I*frac(r)
>>> frac(1 + r*I)
frac(I*r + 1)    # instead of I*frac(r)
```

leosartaj commented:
I have improved (or tried) to for complex arguments. Have a look @jksuom

jksuom commented:
I think it is a good idea to separate the real and imaginary parts. But what happens if ``real`` or ``imag`` has several terms not all of which are numbers (e.g. ``frac(n + 0.5)``)? It might be better to first collect the terms in two lists ``reals`` and ``imags`` in order to apply ``_eval`` to each term separately. And perhaps ``_eval`` could return ``cls(arg, evaluate=False)`` instead of ``None``,

It could also be possible to ``_eval`` each term already in the for loop.

leosartaj commented:
> I think it is a good idea to separate the real and imaginary parts. But what happens if real or imag has several terms not all of which are numbers (e.g. frac(n + 0.5))? 

It will be left unevaluated, terms will be broken to imaginary and real part. Real part consists of explicitly real elements and those whose value could not be determined.

Reason
```
>>> frac(x + 0.5)
frac(x) + 0.5
```
this can pose some issues, if x = 0.5, the answer will be 1 (incorrect)

> It might be better to first collect the terms in two lists reals and imag in order to apply _eval to each term separately. 

By the above argument, I think it's best not to.

> And perhaps _eval could return cls(arg, evaluate=False) instead of None, It could also be possible to _eval each term already in the for loop.

Both are good ideas. Thanks!

jksuom commented:
I probably have an older version for this is what I got:
```>>> x = Symbol('x')
>>> frac(x + 0.5)
frac(x + 0.5)
```
But ``frac(x) + 0.5`` will be ok.

leosartaj commented:
Oh, sorry. I was just explaining the case if it breaks up. You probably have the right version. My point being ``frac(x + 0.5)`` shouldn't break up. As if x is say 0.5, it will evaluate to 1.

jksuom commented:
Yes, you are right. One should not separate ``x`` from 0.5 as ``frac`` is not additive.

Only the real and imaginary parts can be separated safely (I think..). Perhaps one could even go as far as taking ``re(x)`` and ``im(x)`` which is what ``x.as_real_imag()`` gives. This would lead to ``frac(re(x) + 0.5) + I*frac(im(x))``. But is it worse than ``frac(x + 0.5)``?

leosartaj commented:
> Only the real and imaginary parts can be separated safely (I think..)

I agree.

>Only the real and imaginary parts can be separated safely (I think..). Perhaps one could even go as far as taking re(x) and im(x) which is what x.as_real_imag() gives. This would lead to frac(re(x) + 0.5) + I*frac(im(x)). But is it worse than frac(x + 0.5)?

I was thinking on something similar. But in my opinion this result makes it more complex and is of no great use. I think it will be better to leave it unevaluated.

jksuom commented:
> I think it will be better to leave it unevaluated.

I tend to agree. This would probably mean that ``frac`` of an expression containing a symbol will be often left unevaluated. Only if we know that the real and imaginary parts can be separated, we can go further. For example, if ``x = Symbol('x', real=True)``, then ``frac(I*x + 0.5) == I*frac(x)``.

leosartaj commented:
This case is already handled. I think I have added a similar test case as well.

```
>>> x = Symbol('x', real=True)
>>> frac(I*x + 0.5)
0.5 + I*frac(x)
```

jksuom commented:
> ``0.5 + I*frac(x)``

I think 0.5 should not be included.

leosartaj commented:
I don't really see why.

``frac(I*x + 0.5)`` -> ``frac(0.5) + I*frac(x)`` -> ``0.5 + I*frac(x)``

jksuom commented:
Sorry, I had ``floor`` in mind.

leosartaj commented:
Let me know if you have any more concerns.
On 5 Jun 2015 18:50, "Kalevi Suominen" <notifications@github.com> wrote:

> Sorry, I had floor in mind.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/9342#issuecomment-109291152>.
>

jksuom commented:
I think it is beginning to look like a final version. Let's wait for the tests to finish. And then, if there are no objections, you could squash the commits.

leosartaj committed:
Implemented fractional part function

Represents the fractional part of x

For real numbers it is defined [1]_ as

.. math::
    x - \lfloor{x}\rfloor

Examples
========

```
>>> from sympy import Symbol, frac, Rational, floor, ceiling, I
>>> frac(Rational(4, 3))
1/3
>>> frac(-Rational(4, 3))
2/3
```

returns zero for integer arguments

```
>>> n = Symbol('n', integer=True)
>>> frac(n)
0
```

rewrite as floor

```
>>> x = Symbol('x')
>>> frac(x).rewrite(floor)
x - floor(x)
```

for complex arguments

```
>>> r = Symbol('r', real=True)
>>> t = Symbol('t', real=True)
>>> frac(t + I*r)
I*frac(r) + frac(t)
```

See Also
========

sympy.functions.elementary.integers.floor
sympy.functions.elementary.integers.ceiling

References
===========

.. [1] http://en.wikipedia.org/wiki/Fractional_part
.. [2] http://mathworld.wolfram.com/FractionalPart.html

jksuom commented:
Thank you. Merging in a few hours.
